sktree/stats/forestht.py

```diff
@@ -0,0 +1,862 @@
+        (n_estimators, n_samples_final)
+        The predicted posterior probabilities of the samples used in the final test.
+
+    null_dist_ : ArrayLike of shape (n_repeats,)
+        The null distribution of the test statistic.
+
+    Notes
+    -----
+    This class trains two forests: one on the original dataset, and one on the
+    permuted dataset. The forest from the original dataset is cached and re-used to
+    compute the test-statistic each time the :meth:`test` method is called. However,
+    the forest from the permuted dataset is re-trained each time the :meth:`test` is called
+    if the ``covariate_index`` differs from the previous run.
+
+    To fully start from a new dataset, call the ``reset`` method, which will then
+    re-train both forests upon calling the :meth:`test` and :meth:`statistic` methods.
+
+    References
+    ----------
+    .. footbibliography::
+    """
+
+    def __init__(
+        self,
+        estimator=None,
+        random_state=None,
+        verbose=0,
+        test_size=0.2,
+        permute_per_tree=True,
+        sample_dataset_per_tree=True,
+    ):
+        super().__init__(
+            estimator=estimator,
+            random_state=random_state,
+            verbose=verbose,
+            test_size=test_size,
+            permute_per_tree=permute_per_tree,
+            sample_dataset_per_tree=sample_dataset_per_tree,
+        )
+
+    def _get_estimator(self):
+        if self.estimator is None:
+            estimator_ = RandomForestClassifier()
+        elif not isinstance(self.estimator, (ForestClassifier, sklearnForestClassifier)):
+            raise RuntimeError(f"Estimator must be a ForestClassifier, got {type(self.estimator)}")
+        else:
+            # self.estimator is an instance of a ForestEstimator
+            estimator_ = self.estimator
+        return estimator_
+
+    def _statistic(
+        self,
+        estimator: ForestClassifier,
+        X: ArrayLike,
+        y: ArrayLike,
+        covariate_index: ArrayLike,
+        metric: str,
+        return_posteriors: bool,
+        **metric_kwargs,
+    ):
+        """Helper function to compute the test statistic."""
+        metric_func: Callable[[ArrayLike, ArrayLike], float] = METRIC_FUNCTIONS[metric]
+        rng = np.random.default_rng(self.random_state)
+
+        if metric in POSTERIOR_FUNCTIONS:
+            predict_posteriors = True
+        else:
+            predict_posteriors = False
+
+        if predict_posteriors:
+            # now initialize posterior array as (n_trees, n_samples_test, n_classes)
+            posterior_arr = np.full(
+                (self.n_estimators, self._n_samples_, estimator.n_classes_), np.nan
+            )
+        else:
+            # now initialize posterior array as (n_trees, n_samples_test, n_outputs)
+            posterior_arr = np.full(
+                (self.n_estimators, self._n_samples_, estimator.n_outputs_), np.nan
+            )
```

Comment:
@adam2392 I don't see why the statistic metric would affect the shape of posteriors, and the first case does not cover multi-output situations. The shape should be like: 
```python
(self.n_estimators, self._n_samples_, estimator.n_outputs_, estimator.n_classes_[0])
```

Reply:
If we don't predict the posteriors, we simply predict the class (i.e. via predict).

If we predict the posteriors, you're right this does not support multi-output situations right now, but I presumed that was fine? 

> (self.n_estimators, self._n_samples_, estimator.n_outputs_, estimator.n_classes_[0])

Would not be sufficient. Each output may have a different number of classes, so the strategy in scikit-learn is to then return a list of arrays instead, since the last dimension can vary.